### DISCUSSION TITLE
Events IPC

### DISCUSSION THREAD

**Discussion by mentalblock** [YaLTeR/niri#411] [Ideas]
I would like to suggest a separate, read-only Events IPC socket. This is something that the [Hyprland compositor provides](https://github.com/hyprwm/hyprland-wiki/blob/c56516109eea6dabd64d99c03ae4b0869a2612d4/pages/IPC/_index.md#xdg_runtime_dirhyprhissocket2sock).

All events can be monitored by watching this socket. This would allow bars that have no plans to implement Wayland protocols to retrieve status changes immediately and efficiently. In particular, [Eww](https://github.com/elkowar/eww) is the bar I currently use. It has [a listening mechanism](https://elkowar.github.io/eww/configuration.html), `deflisten`, that monitors the output of a command and updates the bar UI.

In summary, if Niri had a read-only Events IPC, I would then be able to use the `socat` utility to extract and parse whatever Niri state I would like. It is lightweight and responsive method. My only use case at the moment is for workspaces status and focused window titles, but I am sure others would leverage this to suit their needs.

Related discussions and issues:
* https://github.com/YaLTeR/niri/discussions/209
* https://github.com/YaLTeR/niri/issues/135

**Comment by YaLTeR** [comment]
Why a separate socket? My plan is to just make a `niri msg event-stream` (or something) that sends you events indefinitely.

**Reply by mentalblock** [reply]
My apologies. I was not aware of your plan to implement `niri msg event-stream`. My cursory search through the issues and discussions before I started this topic did not indicate anything of the like. I think your planned solution would suffice.

As for whether there are any benefits from a separate IPC socket is unknown to me.

Thanks for the quick response and your work on this project.

**Reply by kls1991** [reply]
> Why a separate socket? My plan is to just make a `niri msg event-stream` (or something) that sends you events indefinitely.

I'm trying to have some application-aware keymaps. That `niri msg event-stream` will be a life saver.

**Reply by morenathan** [reply]
I'll third this!

In another life, my favorite Window Manager was WMII, which used the same principle for absolutely everything in the window manager.

The whole keyboard and mouse interface ran on a loop that was fed by the WM. You simply looked for the events that suited your needs and issued the appropriate commands back.

(EDIT: The coolest this about this? You could write that loop in any language that could open a file.)

WMII and Niri are very much the same, with the exception that WMII was panoramic, meaning you had from the far left edge of your screens to the right, for all your windowing. Along with tags and a floating layer, it was my go-to WM.

**Reply by nebulosa2007** [reply]
Could be related to this conversation.. 

I want to make one more `waybar` module - language module. After searching for info about showing active layout I realized that in Wayland only Niri itself should provide info about active keyboard layout  .

Niri already has an action:
`niri msg action switch-layout next|prev` but nothing is displayed after it.

**Reply by morenathan** [reply]
@nebulosa2007 Same issue I have with almost all actions. I'm not really sure how the current event system is of any use without communication.

**Reply by mentalblock** [reply]
> Why a separate socket? My plan is to just make a `niri msg event-stream` (or something) that sends you events indefinitely.

A possible downside of running a separate process is that it may seem unnecessary friction when the alternative could be to just open a file (socket) when programmatically trying to interact with Niri

**Reply by YaLTeR** [reply]
All `niri msg` is doing is opening a socket and writing a json message. No problem doing that manually.

**Reply by mentalblock** [reply]
I am not sure I understand what you mean. In the context of a separate, read-only socket I proposed here, you asked if there was any reason for such a proposal if you were already creating something like `niri msg event-stream`. At first I could not think of anything.

I then realized that if someone wanted to programmatically monitor events from Niri, running a separate process such as `niri msg event-stream` may be more cumbersome than reading a file (socket). This is all I meant. When you reply about writing a JSON message, I feel that we have maybe misunderstood each other as monitoring events from a separate Events IPC socket would have nothing to do with writing JSON messages.

**Reply by YaLTeR** [reply]
`niri msg event-stream` is (will be) equivalent to connecting to `$NIRI_SOCKET`, writing a json message there, then just reading events. Which is not that much harder than connecting to a (second proposed) socket and reading events. There's only one extra step (writing a json message) of difference between the two.

**Comment by YaLTeR** [comment]
It's implemented now: https://github.com/YaLTeR/niri/wiki/IPC#event-stream

**Reply by salahoued** [reply]
Hi, I just started using `niri`, and I use `eww` as a status/widgets bar, I have made my `eww` config to work with `hyprland` ans `labwc` to show the workspaces, active windows and keyboard layout (using 2 layouts), but I still can't wrap my head around the way to make it work with `niri`.

this is how I get workspaces on `hyprland` (from its wiki)
```
#!/usr/bin/env bash

spaces (){
  WORKSPACE_WINDOWS=$(hyprctl workspaces -j | jq 'map({key: .id | tostring, value: .windows}) | from_entries')
  WORKSPACE_NUMBER=$(echo "$WORKSPACE_WINDOWS" | jq '. | length')
  seq 1 "$WORKSPACE_NUMBER" | jq --argjson windows "${WORKSPACE_WINDOWS}" --slurp -Mc 'map(tostring) | map({id: ., windows: ($windows[.]//0)})'
}

spaces
socat -u UNIX-CONNECT:$XDG_RUNTIME_DIR/hypr/$HYPRLAND_INSTANCE_SIGNATURE/.socket2.sock - | while read -r line; do
  spaces
done
```
is there any way to adapt it to work with `niri`

Thanks for all your hard work.

**Reply by YaLTeR** [reply]
Everything you mentioned is possible (Waybar implements it for example). This discussion has a bunch of examples for workspaces: https://github.com/YaLTeR/niri/discussions/356

**Comment by roceb** [comment]
You can use niri msg windows | grep windows-id | wc -l
